### DISCUSSION TITLE
Understanding the non-cudf kernels in Nsight System Profling results

### DISCUSSION THREAD

**Discussion by fandebiao01** [NVIDIA/spark-rapids#11797] [Q&A] [answered]
Hi,
I used Nsight sys to profile the second query in https://github.com/NVIDIA/spark-rapids-examples/blob/main/examples/SQL%2BDF-Examples/micro-benchmarks/notebooks/micro-benchmarks-gpu.ipynb

In the result I'm seeing some non-cuda kernels, especially the **DeviceMergeSort** and **static_kernel**, their function name doesn't has the cuda perfix, so I have no idea how to find the source of them. If I zommed into the desription, I can see some cudf name though the whole description doesn't quite make sense as shown in the second screenshot.
what are they used for? and where can I find the source of them? Thanks.

![image](https://github.com/user-attachments/assets/efd1f483-3006-4e1d-bef3-171a4e70baf5)

![image](https://github.com/user-attachments/assets/c84b127f-e48a-4434-94fb-b6ff38a9a469)

**Comment by jlowe** [comment, marked as answer]
DeviceMergeSort and static_kernel typically groups kernels associated with CUB, whereas _kernel_agent typically groups kernels from Thrust.  CUB and Thrust are tools leveraged by libcudf and spark-rapids-jni.  See https://nvidia.github.io/cccl/cub/index.html and https://nvidia.github.io/cccl/thrust/, respectively.  CUB and Thrust are part of CUDA Core Compute Libraries (CCCL), see https://github.com/NVIDIA/cccl.

**Reply by fandebiao01** [reply]
Thanks for the clear and prompt anwser!
